scalagen/src/main/scala/org/scalameta/scalagen/Runner.scala

```diff
@@ -0,0 +1,329 @@
+      case c: CompanionGenerator => applyCompanionExtender(in, c)
+      case t: TransmutationGenerator => applyTransmutator(in, t)
+    }
+  }
+
+  private def applyExtender[A <: Tree](in: A, g: ExtensionGenerator): A = {
+    val res = in match {
+      case c: Defn.Class =>
+        val clazz = c.withStats(c.extract[Stat] ::: g.extend(c))
+        removeAnnot(clazz, g.name)
+      case o: Defn.Object =>
+        val obj = o.withStats(o.extract[Stat] ::: g.extend(o))
+        removeAnnot(obj, g.name)
+      case t: Defn.Trait =>
+        val trat = t.withStats(t.extract[Stat] ::: g.extend(t))
+        removeAnnot(trat, g.name)
+    }
+
+    res.asInstanceOf[A]
+  }
+
+  private def genCompanion(name: Term.Name, stats: List[Stat]): Defn.Object =
+    q"object $name { ..$stats }"
+
+  private def applyCompanionExtender[A <: Tree](in: A, g: CompanionGenerator): A = {
+    assert(
+      in.parent.isDefined,
+      s"Companion generator '${g.name}' relies on the annotee having a parent")
+
+    val res = in match {
+      case c: Defn.Class =>
+        val stats = g.extendCompanion(c)
+        val clazzWithoutAnnot: Defn.Class = removeAnnot(c, g.name)
+        if (c.companionObject.isEmpty) {
+          val defns: List[Defn] = clazzWithoutAnnot :: genCompanion(c.name.asTerm, stats) :: Nil
+          transmutationCache.put(
+            Structurally(withoutStats(c.parent.get)),
+            TransmutationResult(clazzWithoutAnnot, defns))
+        } else {
+          //TODO: Check that the companion has not already been transformed
+          companionExtensionCache.put(c.name.value, stats)
+        }
+        clazzWithoutAnnot
+      case t: Defn.Type =>
+        val stats = g.extendCompanion(t)
+        val typeWithoutAnnot: Defn.Type = removeAnnot(t, g.name)
+        if (t.companionObject.isEmpty) {
+          val defns: List[Defn] = typeWithoutAnnot :: genCompanion(t.name.asTerm, stats) :: Nil
+          transmutationCache.put(
+            Structurally(withoutStats(t.parent.get)),
+            TransmutationResult(typeWithoutAnnot, defns))
+        } else {
+          //TODO: Check that the companion has not already been transformed
+          companionExtensionCache.put(t.name.value, stats)
+        }
+        typeWithoutAnnot
+      case t: Defn.Trait =>
+        val stats = g.extendCompanion(t)
+        val traitWithoutAnnot: Defn.Trait = removeAnnot(t, g.name)
+        if (t.companionObject.isEmpty) {
+          val defns: List[Defn] = traitWithoutAnnot :: genCompanion(t.name.asTerm, stats) :: Nil
+          transmutationCache.put(
+            Structurally(withoutStats(t.parent.get)),
+            TransmutationResult(traitWithoutAnnot, defns))
+        } else {
+          //TODO: Check that the companion has not already been transformed
+          companionExtensionCache.put(t.name.value, stats)
+        }
+        traitWithoutAnnot
+    }
+
+    res.asInstanceOf[A]
+  }
+
+  // TODO: This method should not be necessary
```

Comment:
Why is this method necessary? I'm not sure why we cache by the structure of this tree with stats removed.

Reply:
It's a horrible hack. I'll try and find a way around it.

Basically, when a node's children are being transmuted, they set aside a list of definitions their parent need to add...

We need to find the child later (when we are transforming its parent), in order to remove that child and add the defs as children. Due to the nature of transforms Tree => Tree signature, we cannot remove the Child as part of its transformation.

We have to store the parent without stats as when we transform the parent. It will be different to the input parent, as it's children will have changed, thus structural equality fails.

It would be nice to get around this somehow, I just do not have a good idea. Unless we use FQN's and a semantic database.

Reply:
> Unless we use FQN's and a semantic database.

We can get FQNs for definitions without running compile. We do that in language-server to support "Go to definition" on the classpath for both java and scala dependencies (that have no semanticdb). The implementation is quite simple 

https://github.com/scalameta/language-server/blob/master/metaserver/src/main/scala/scala/meta/languageserver/mtags/ScalaMtags.scala
https://github.com/scalameta/language-server/blob/master/metaserver/src/test/scala/tests/mtags/ScalaMtagsTest.scala